### PR TITLE
Fb authentification

### PR DESCRIPTION
**Pull request by RobPavl**
Made opportunity authenticate via Facebook

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -34,13 +34,14 @@ gem 'slim'
...
-gem 'therubyracer'
-gem 'omniauth'
-gem 'omniauth-facebook'
+
+  gem 'devise', '~> 3.4.1'
+  gem 'bootstrap-sass', '~> 3.3.1.0'
+  gem 'devise-bootstrap-views', '~> 0.0.3'
+  gem 'execjs'
+  gem 'therubyracer'
+  gem 'omniauth-facebook'
```
Comment (pyotrkoryakin) [inline Gemfile]: fix indentation here. 
follow https://github.com/arbox/ruby-style-guide/blob/master/README-ruRU.md

File: app/assets/stylesheets/callbacks.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Callbacks controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (pyotrkoryakin) [inline app/assets/stylesheets/callbacks.scss]: Please, Remove this file at all.

File: app/assets/stylesheets/sessions.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Sessions controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (pyotrkoryakin) [inline app/assets/stylesheets/sessions.scss]: Please, Remove this file at all.

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -5,6 +5,7 @@ class ApplicationController < ActionController::Base
 
    before_action :configure_permitted_parameters, if: :devise_controller?
 
+
```
Comment (pyotrkoryakin) [inline app/controllers/application_controller.rb]: Why this newline had appeared here? It is important to keep version history clean and readable. Such random lines makes it hard to understand what changes had been made to particular file.

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+def facebook
+     @user = User.from_omniauth(request.env["omniauth.auth"])
+     if @user.persisted?
+      sign_in_and_redirect @user, :event => :authentication #this will throw if @user is not activated
+      set_flash_message(:notice, :success, :kind => "Facebook") if is_navigational_format?
+    else
+      session["devise.facebook_data"] = request.env["omniauth.auth"]
+      redirect_to new_user_registration_url
+    end
+  end
```
Comment (pyotrkoryakin) [inline app/controllers/users/omniauth_callbacks_controller.rb]: fix indentation here. 
follow https://github.com/arbox/ruby-style-guide/blob/master/README-ruRU.md

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -5,10 +5,20 @@ def index
     @users = User.all
   end
 
+  def new
+    @add_user = User.new
+  end
```
Comment (pyotrkoryakin) [inline app/controllers/users_controller.rb]: Is this method used anywhere? If it is not? you should remove it.

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -5,10 +5,20 @@ def index
...
     @user = User.find(params[:id])
     unless @user == current_user
       redirect_to :back, :alert => "Access denied."
     end
   end
+
+  def destroy
+    session[:user_id] = nil
+    redirect_to root_url
+  end
```
Comment (pyotrkoryakin) [inline app/controllers/users_controller.rb]: Is this method used anywhere? If it is not? you should remove it.

File: app/assets/javascripts/callbacks.coffee [hunk 8]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (pyotrkoryakin) [inline app/assets/javascripts/callbacks.coffee]: Please, Remove this file at all.

File: app/assets/javascripts/sessions.coffee [hunk 9]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (pyotrkoryakin) [inline app/assets/javascripts/sessions.coffee]: Please, Remove this file at all.

File: app/helpers/callbacks_helper.rb [hunk 10]
```diff
@@ -0,0 +1,2 @@
+module CallbacksHelper
+end
```
Comment (pyotrkoryakin) [inline app/helpers/callbacks_helper.rb]: Please, Remove this file at all.

File: app/helpers/sessions_helper.rb [hunk 11]
```diff
@@ -0,0 +1,2 @@
+module SessionsHelper
+end
```
Comment (pyotrkoryakin) [inline app/helpers/sessions_helper.rb]: Please, Remove this file at all.

File: app/models/user.rb [hunk 12]
```diff
@@ -1,7 +1,26 @@
...
+    user.name = auth.info.name
+  end
+end
+
+def self.new_with_session(params, session)
+    super.tap do |user|
+      if data = session["devise.facebook_data"] && session["devise.facebook_data"]["extra"]["raw_info"]
+        user.email = data["email"] if user.email.blank?
+      end
+    end
```
Comment (pyotrkoryakin) [inline app/models/user.rb]: Fix indentation here. 
Our code shhoud follow https://github.com/arbox/ruby-style-guide/blob/master/README-ruRU.md

File: app/views/layouts/application.html.erb [hunk 13]
```diff
@@ -21,5 +21,10 @@
   </div>
 
 
+
+
+
+
+
```
Comment (pyotrkoryakin) [inline app/views/layouts/application.html.erb]: Why these newlines had appeared here? It is important to keep version history clean and readable. Such random lines makes it hard to understand what changes had been made to particular file.

File: app/views/static_pages/landing.html.slim [hunk 14]
```diff
@@ -1 +1,6 @@
 h1= @greeting
+
+
+
+= link_to 'Page sample'
+
```
Comment (pyotrkoryakin) [inline app/views/static_pages/landing.html.slim]: This link is useless here. We had moved such stuff to navigation bar. I think it appeared here again as the result of rebase. Please, remove it

File: app/views/static_pages/petro.html.slim [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+
+a href='https://github.com/pyotrkoryakin' target='blank'
+  'This is My GitHub Page
+
+/ <h1>
+/   <%= "This is #{@name}'s page" %>
+/ </h1>
+/ <a href="https://github.com/pyotrkoryakin" target="blank">
+/   This is My GitHub Page
+/ </a>
```
Comment (pyotrkoryakin) [inline app/views/static_pages/petro.html.slim]: This file was also removed. You should not keep it.

File: config/routes.rb [hunk 16]
```diff
@@ -15,14 +15,6 @@
...
   #   get 'products/:id' => 'catalog#view'
 
- #   match 'auth/:provider/callback', to: 'sessions#create', via: [:get, :post]
- #   match 'auth/failure', to: redirect('/'), via: [:get, :post]
-  #  match 'signout', to: 'sessions#destroy', as: 'signout', via: [:get, :post]
-
-#get 'auth/:provider', to: 'sessions#create'
-#get 'auth/failure', to: redirect('/')
-#get 'signout', to: 'sessions#destroy', as: 'signout'
-
```
Comment (pyotrkoryakin) [inline config/routes.rb]: It is a good practice to remove commented code (yes)

File: spec/helpers/callbacks_helper_spec.rb [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+# describe CallbacksHelper do
+#   describe "string concat" do
+#     it "concats two strings with spaces" do
+#       expect(helper.concat_strings("this","that")).to eq("this that")
+#     end
+#   end
+# end
+RSpec.describe CallbacksHelper, :type => :helper do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (pyotrkoryakin) [inline spec/helpers/callbacks_helper_spec.rb]: Please, remove this file. We do not have callbacks helpers

File: spec/helpers/sessions_helper_spec.rb [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+# describe SessionsHelper do
+#   describe "string concat" do
+#     it "concats two strings with spaces" do
+#       expect(helper.concat_strings("this","that")).to eq("this that")
+#     end
+#   end
+# end
+RSpec.describe SessionsHelper, :type => :helper do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (pyotrkoryakin) [inline spec/helpers/sessions_helper_spec.rb]: Please, remove this file. We do not have sessions helpers

File: app/assets/javascripts/sessions.coffee (hunk 9 above)
Comment (RobPavl) [inline app/assets/javascripts/sessions.coffee]: completed

File: app/assets/javascripts/callbacks.coffee (hunk 8 above)
Comment (RobPavl) [inline app/assets/javascripts/callbacks.coffee]: completed

File: app/assets/stylesheets/callbacks.scss (hunk 2 above)
Comment (RobPavl) [inline app/assets/stylesheets/callbacks.scss]: completed

File: app/assets/stylesheets/sessions.scss (hunk 3 above)
Comment (RobPavl) [inline app/assets/stylesheets/sessions.scss]: completed

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (RobPavl) [inline app/controllers/application_controller.rb]: Have deleted this newline

File: app/helpers/callbacks_helper.rb (hunk 10 above)
Comment (RobPavl) [inline app/helpers/callbacks_helper.rb]: completed

File: app/helpers/sessions_helper.rb (hunk 11 above)
Comment (RobPavl) [inline app/helpers/sessions_helper.rb]: completed

File: app/views/layouts/application.html.erb (hunk 13 above)
Comment (RobPavl) [inline app/views/layouts/application.html.erb]: Have deleted this newlines

File: app/views/static_pages/landing.html.slim (hunk 14 above)
Comment (RobPavl) [inline app/views/static_pages/landing.html.slim]: Have deleted this link

File: app/views/static_pages/petro.html.slim (hunk 15 above)
Comment (RobPavl) [inline app/views/static_pages/petro.html.slim]: have delete it

File: config/routes.rb (hunk 16 above)
Comment (RobPavl) [inline config/routes.rb]: completed

File: spec/helpers/callbacks_helper_spec.rb (hunk 17 above)
Comment (RobPavl) [inline spec/helpers/callbacks_helper_spec.rb]: completed

File: spec/helpers/sessions_helper_spec.rb (hunk 18 above)
Comment (RobPavl) [inline spec/helpers/sessions_helper_spec.rb]: completed

File: app/models/user.rb (hunk 12 above)
Comment (RobPavl) [inline app/models/user.rb]: completed

File: Gemfile (hunk 1 above)
Comment (RobPavl) [inline Gemfile]: compleed

File: spec/helpers/sessions_helper_spec.rb (hunk 18 above)
Comment (pyotrkoryakin) [inline spec/helpers/sessions_helper_spec.rb]: @RobPavl , when I receive mail notification about your 'completed' comment, I expect that when you say 'completed' on particular issue it is really completed. Issue is Completed when it is  fixed, tested, committed and pushed to github.
But now I'm confused - what does this comment mean?

File: app/controllers/users/omniauth_callbacks_controller.rb (hunk 5 above)
Comment (RobPavl) [inline app/controllers/users/omniauth_callbacks_controller.rb]: completed

**Comment by pyotrkoryakin** [comment]
Now it looks great! :+1:

**Pull request merged**